File: TodayPlan.swift
//
//  TodayPlan.swift
//  Avenue3
//

import Foundation

/// A spontaneous one-on-one activity post scoped to the current calendar day.
/// Any user can post an open plan; any other user can claim it first-come-first-served.
/// Once claimed, both participants get a chat thread (via the existing messaging layer)
/// and are each asked independently after the scheduled time whether the other showed up.
struct TodayPlan: Identifiable, Codable {
    let id: String
    let creatorID: String
    let activity: Activity
    /// Specific date and time the hangout is planned for (must fall on today's calendar date)
    let scheduledTime: Date
    /// Optional short note (e.g. "bring sunscreen") — kept separate from location.
    let note: String?
    /// Where the plan is happening — free text, required at creation via PlanLocationField.
    /// Optional here (not at the UI layer) so plans posted before this field existed still decode.
    let location: String?
    /// Name of the place selected via location search (nil for free-typed locations).
    let locationName: String?
    /// Coordinates of the place selected via location search, for tap-to-navigate.
    let locationLatitude: Double?
    let locationLongitude: Double?
    var status: TodayPlanStatus
    /// Set to the claiming user's ID the moment someone claims the plan
    var claimerID: String?
    let createdAt: Date
    var updatedAt: Date

    // MARK: - Show-up Reports
    // Each participant independently reports on the other after scheduledTime passes.
    // Mirrors Match.user1Decision / user2Decision — optional Bool, nil means not yet reported.
    // A nil never counts against either side.

    /// Creator's thumbs-up (true) or thumbs-down (false) about the claimer; nil until submitted
    var creatorReportedClaimer: Bool?
    /// Claimer's thumbs-up (true) or thumbs-down (false) about the creator; nil until submitted
    var claimerReportedCreator: Bool?

    init(
        id: String = UUID().uuidString,
        creatorID: String,
        activity: Activity,
        scheduledTime: Date,
        note: String? = nil,
        location: String? = nil,
        locationName: String? = nil,
        locationLatitude: Double? = nil,
        locationLongitude: Double? = nil,
        status: TodayPlanStatus = .open,
        claimerID: String? = nil,
        createdAt: Date = Date(),
        updatedAt: Date = Date(),
        creatorReportedClaimer: Bool? = nil,
        claimerReportedCreator: Bool? = nil
    ) {
        self.id = id
        self.creatorID = creatorID
        self.activity = activity
        self.scheduledTime = scheduledTime
        self.note = note
        self.location = location
        self.locationName = locationName
        self.locationLatitude = locationLatitude
        self.locationLongitude = locationLongitude
        self.status = status
        self.claimerID = claimerID
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.creatorReportedClaimer = creatorReportedClaimer
        self.claimerReportedCreator = claimerReportedCreator
    }

    // MARK: - Derived State

    /// True when scheduledTime is in the past. No stored flag — derived on every read.
    var isExpired: Bool {
        scheduledTime < Date()
    }

    /// Returns the other participant's user ID (creator ↔ claimer).
    func otherUserID(for userID: String) -> String? {
        if userID == creatorID { return claimerID }
        if userID == claimerID { return creatorID }
        return nil
    }

    /// True when this user still needs to submit a show-up report about the other person.
    func awaitingReport(from userID: String) -> Bool {
        guard isExpired, status == .claimed, claimerID != nil else { return false }
        if userID == creatorID { return creatorReportedClaimer == nil }
        if userID == claimerID { return claimerReportedCreator == nil }
        return false
    }

    /// The thumbs-up/down this user submitted about the other person, or nil if not yet reported.
    func report(givenBy userID: String) -> Bool? {
        if userID == creatorID { return creatorReportedClaimer }
        if userID == claimerID { return claimerReportedCreator }
        return nil
    }
}

// MARK: - Status

enum TodayPlanStatus: String, Codable, CaseIterable {
    case open = "open"
    case claimed = "claimed"

    var displayName: String {
        switch self {
        case .open:    "Open"
        case .claimed: "Claimed"
        }
    }
}

File: android/app/src/main/java/com/georgeappdev/atxfriends/data/model/TodayPlan.kt
package com.georgeappdev.atxfriends.data.model

import com.georgeappdev.atxfriends.data.firestore.DocReader
import com.georgeappdev.atxfriends.data.firestore.TodayPlanFields
import java.time.Instant

/** `todayPlans/{planID}` — an open slot on the Today tab that someone can claim. */
data class TodayPlan(
    val id: String,
    val creatorID: String,
    val activity: Activity,
    val scheduledTime: Instant,
    val note: String?,
    val location: String?,
    val locationName: String?,
    val locationLatitude: Double?,
    val locationLongitude: Double?,
    val status: TodayPlanStatus,
    val claimerID: String?,
    val createdAt: Instant,
    val updatedAt: Instant,
    /** Show-up verdicts, written server-side by the applyShowUpReport Cloud Function. */
    val creatorReportedClaimer: Boolean?,
    val claimerReportedCreator: Boolean?,
) {
    companion object {
        /**
         * Same required fields as iOS `TodayPlanService.decode`, except an unrecognized
         * `status` decodes to [TodayPlanStatus.UNKNOWN] instead of dropping the plan.
         */
        fun fromFirestore(id: String, data: Map<String, Any?>): TodayPlan? {
            val r = DocReader(data)
            return TodayPlan(
                id = id,
                creatorID = r.string(TodayPlanFields.CREATOR_ID) ?: return null,
                activity = Activity.fromFirestore(r.map(TodayPlanFields.ACTIVITY)) ?: return null,
                scheduledTime = r.instant(TodayPlanFields.SCHEDULED_TIME) ?: return null,
                status = TodayPlanStatus.fromRaw(r.string(TodayPlanFields.STATUS) ?: return null),
                createdAt = r.instant(TodayPlanFields.CREATED_AT) ?: return null,
                updatedAt = r.instant(TodayPlanFields.UPDATED_AT) ?: return null,
                note = r.string(TodayPlanFields.NOTE),
                location = r.string(TodayPlanFields.LOCATION),
                locationName = r.string(TodayPlanFields.LOCATION_NAME),
                locationLatitude = r.double(TodayPlanFields.LOCATION_LATITUDE),
                locationLongitude = r.double(TodayPlanFields.LOCATION_LONGITUDE),
                claimerID = r.string(TodayPlanFields.CLAIMER_ID),
                creatorReportedClaimer = r.bool(TodayPlanFields.CREATOR_REPORTED_CLAIMER),
                claimerReportedCreator = r.bool(TodayPlanFields.CLAIMER_REPORTED_CREATOR),
            )
        }
    }
}
